source/SupportLib/PlayGamesPluginSupport/src/main/java/com/google/games/bridge/RequestPermissionsRequest.java

```diff
@@ -0,0 +1,108 @@
+    private Scope[] scopes;
+    private HelperFragment helperFragment;
+
+    public RequestPermissionsRequest(Scope[] scopes) {
+        this.scopes = scopes;
+    }
+
+    Task<Result> getTask() {
+        return resultTaskSource.getTask();
+    }
+
+    public void process(HelperFragment helperFragment) {
+        this.helperFragment = helperFragment;
+        final Activity activity = helperFragment.getActivity();
+        GoogleSignInAccount account = HelperFragment.getAccount(activity);
+        Scope[] unauthorizedScopes = getUnauthorizedScopes(account, scopes);
+        if (unauthorizedScopes.length == 0) {
+            setSuccess(GoogleSignIn.getAccountForScopes(activity, scopes[0], scopes));
+        } else {
+            Intent intent = getSignInIntentForAccountAndScopes(activity, account, unauthorizedScopes);
+            helperFragment.startActivityForResult(intent, HelperFragment.RC_SHOW_REQUEST_PERMISSIONS_UI);
+        }
+    }
+
+    private Scope[] getUnauthorizedScopes(GoogleSignInAccount account, Scope[] scopes) {
+        ArrayList<Scope> unauthorizedScopes = new ArrayList<Scope>();
+        for (Scope scope : scopes) {
+            if (!GoogleSignIn.hasPermissions(account, scope)) {
+                unauthorizedScopes.add(scope);
+            }
+        }
+        return unauthorizedScopes.toArray(new Scope[unauthorizedScopes.size()]);
+    }
+
+    private static Intent getSignInIntentForAccountAndScopes(
+        /* @NonNull */ Activity activity, /* @Nullable */ GoogleSignInAccount account, /* @NonNull */ Scope... scopes) {
+        GoogleSignInOptions.Builder optionsBuilder = new GoogleSignInOptions.Builder();
+
+        if (scopes.length > 0) {
+            optionsBuilder.requestScopes(scopes[0], scopes);
+        }
+
+        if (account != null && account.getEmail() != null) {
+            optionsBuilder.setAccountName(account.getEmail());
+        }
+
+        return GoogleSignIn.getClient(activity, optionsBuilder.build()).getSignInIntent();
+    }
+
+    public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == HelperFragment.RC_SHOW_REQUEST_PERMISSIONS_UI) {
+            GoogleSignInResult result = Auth.GoogleSignInApi.getSignInResultFromIntent(data);
+            if (result != null && result.isSuccess()) {
+                GoogleSignInAccount account = result.getSignInAccount();
+                setSuccess(account);
+            } else if (resultCode == Activity.RESULT_CANCELED) {
+                setResult(CommonStatusCodes.CANCELED);
+            } else if (result != null) {
+                setResult(result.getStatus().getStatusCode());
+            } else {
+                setResult(CommonStatusCodes.ERROR);
+            }
+        }
+    }
+
+    void setResult(int code) {
+        Result result = new Result(code, /* account= */ null);
+        resultTaskSource.setResult(result);
```

Comment:
If code means error it should call setException instead of setResult

Reply:
Made a bit of changes in this file. TaskCompletionSource<Result> is replaced with TaskCompletionSource<GoogleSignInAccount> and now this code returns GoogleSignInAccount if it can. Otherwise, it returns exceptions including failure codes. 

Right now, it sets exeption here and I also renamed the method name.